File: scripts/bmo-site-caretaker.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
from pathlib import Path

DEFAULT_SITE_DIR = Path.home() / "prismtek-site"
DEFAULT_REPLICA_DIR = Path.home() / "prismtek-site-replica"
DEFAULT_OUTPUT = Path("workflows") / "bmo-site-caretaker.json"
DEFAULT_DISCOVERY_ROOT = Path.home()
MAX_DISCOVERY_DEPTH = 4
CHAT_SURFACE_TOKENS = ("chat", "assistant", "api", "worker", "function")


def within_depth(base: Path, candidate: Path, max_depth: int) -> bool:
    try:
        rel = candidate.relative_to(base)
    except ValueError:
        return False
    return len(rel.parts) <= max_depth


def discover_repo(root: Path, name: str) -> list[str]:
    if not root.exists():
        return []
    matches: list[str] = []
    for path in root.rglob(name):
        if path.is_dir() and within_depth(root, path, MAX_DISCOVERY_DEPTH):
            matches.append(str(path))
    return sorted(set(matches))[:20]


def scan_site(root: Path) -> dict[str, object]:
    html_files = []
    asset_count = 0
    chat_candidates = []
    if not root.exists():
        return {
            "path": str(root),
            "exists": False,
            "html_files": [],
            "asset_count": 0,
            "chat_surface_candidates": [],
        }

    for path in root.rglob("*"):
        if path.is_file():
            suffix = path.suffix.lower()
            rel = path.relative_to(root).as_posix()
            if suffix in {".html", ".htm"}:
                html_files.append(rel)
            if suffix in {".png", ".jpg", ".jpeg", ".gif", ".svg", ".webp", ".css", ".js"}:
                asset_count += 1
            lowered = rel.lower()
            if suffix in {".html", ".htm", ".js", ".ts", ".tsx", ".jsx", ".json", ".md"} and any(
                token in lowered for token in CHAT_SURFACE_TOKENS
            ):
                chat_candidates.append(rel)

    return {
        "path": str(root),
        "exists": True,
        "html_files": sorted(html_files),
        "asset_count": asset_count,
        "chat_surface_candidates": sorted(set(chat_candidates)),
    }


def scan_replica(root: Path) -> dict[str, object]:
    if not root.exists():
        return {
            "path": str(root),
            "exists": False,
            "routes": [],
            "components": [],
            "chat_surface_candidates": [],
        }

    routes = []
    components = []
    chat_candidates = []
    for path in root.rglob("*"):
        if path.is_file() and path.suffix.lower() in {".tsx", ".ts", ".jsx", ".js"}:
            rel = path.relative_to(root).as_posix()
            lowered = rel.lower()
            if any(token in lowered for token in {"route", "page", "app.tsx", "main.tsx"}):
                routes.append(rel)
            if "component" in lowered or "/components/" in lowered:
                components.append(rel)
            if any(token in lowered for token in CHAT_SURFACE_TOKENS):
                chat_candidates.append(rel)

    return {
        "path": str(root),
        "exists": True,
        "routes": sorted(set(routes)),
        "components": sorted(set(components)),
        "chat_surface_candidates": sorted(set(chat_candidates)),
    }


def build_plan(site: dict[str, object], replica: dict[str, object]) -> list[dict[str, object]]:
    plan = []
    for html_file in site.get("html_files", []):
        slug = str(html_file).removesuffix(".html").removesuffix(".htm")
        if slug.endswith("/index"):
            slug = slug[:-6]
        route = "/" if slug in {"", "index"} else f"/{slug.strip('/')}"
        plan.append(
            {
                "route": route,
                "source": html_file,
                "status": "pending-migration",
                "target": "prismtek-site-replica",
            }
        )
    return plan


def main() -> None:
    parser = argparse.ArgumentParser(description="Inventory prismtek-site and prismtek-site-replica for BMO caretaker routing.")
    parser.add_argument("--site-dir", default=os.environ.get("BMO_SITE_DIR", str(DEFAULT_SITE_DIR)))
    parser.add_argument("--replica-dir", default=os.environ.get("BMO_SITE_REPLICA_DIR", str(DEFAULT_REPLICA_DIR)))
    parser.add_argument("--discovery-root", default=os.environ.get("BMO_SITE_DISCOVERY_ROOT", str(DEFAULT_DISCOVERY_ROOT)))
    parser.add_argument("--output", default=str(DEFAULT_OUTPUT))
    args = parser.parse_args()

    site_path = Path(args.site_dir).expanduser()
    replica_path = Path(args.replica_dir).expanduser()
    discovery_root = Path(args.discovery_root).expanduser()

    site = scan_site(site_path)
    replica = scan_replica(replica_path)
    payload = {
        "site": site,
        "replica": replica,
        "discovery": {
            "root": str(discovery_root),
            "site_candidates": [] if site["exists"] else discover_repo(discovery_root, "prismtek-site"),
            "replica_candidates": [] if replica["exists"] else discover_repo(discovery_root, "prismtek-site-replica"),
        },
        "migration_plan": build_plan(site, replica),
        "chat_agent_handoff": {
            "website_owner_repo": "prismtek-site",
            "runtime_contract_repo": "bmo-stack",
            "site_candidates": site.get("chat_surface_candidates", []),
            "replica_candidates": replica.get("chat_surface_candidates", []),
        },
    }

    if not site["exists"] and payload["discovery"]["site_candidates"]:
        payload["site"]["hint"] = "Use --site-dir with one of discovery.site_candidates."
    if not replica["exists"] and payload["discovery"]["replica_candidates"]:
        payload["replica"]["hint"] = "Use --replica-dir with one of discovery.replica_candidates."

    output = Path(args.output)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    print(json.dumps(payload, indent=2))


if __name__ == "__main__":
    main()

File: scripts/bmo-site-caretaker.mjs
#!/usr/bin/env node

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import childProcess from "node:child_process";

const defaultSiteDir = path.join(os.homedir(), "prismtek-site");
const defaultReplicaDir = path.join(os.homedir(), "prismtek-site-replica");
const defaultOutput = path.join("workflows", "bmo-site-caretaker.json");
const defaultDiscoveryRoot = os.homedir();
const maxDiscoveryDepth = 4;
const chatSurfaceTokens = ["chat", "assistant", "api", "worker", "function"];

function parseArgs(argv) {
  const args = {
    siteDir: process.env.BMO_SITE_DIR || defaultSiteDir,
    replicaDir: process.env.BMO_SITE_REPLICA_DIR || defaultReplicaDir,
    discoveryRoot: process.env.BMO_SITE_DISCOVERY_ROOT || defaultDiscoveryRoot,
    output: defaultOutput,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--site-dir") args.siteDir = argv[++index];
    else if (arg === "--replica-dir") args.replicaDir = argv[++index];
    else if (arg === "--discovery-root") args.discoveryRoot = argv[++index];
    else if (arg === "--output") args.output = argv[++index];
    else throw new Error(`unknown argument: ${arg}`);
  }
  return args;
}

function withinDepth(base, candidate, maxDepth) {
  const rel = path.relative(base, candidate);
  if (rel.startsWith("..") || path.isAbsolute(rel)) return false;
  return rel.split(path.sep).filter(Boolean).length <= maxDepth;
}

function discoverRepo(root, name) {
  if (!fs.existsSync(root)) return [];
  const matches = [];
  function walk(current) {
    const entries = fs.readdirSync(current, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      const fullPath = path.join(current, entry.name);
      if (!withinDepth(root, fullPath, maxDiscoveryDepth)) continue;
      if (entry.name === name) matches.push(fullPath);
      walk(fullPath);
    }
  }
  walk(root);
  return [...new Set(matches)].sort().slice(0, 20);
}

function gitTreeFiles(root) {
  const gitDir = path.join(root, ".git");
  if (!fs.existsSync(gitDir)) return null;
  const result = childProcess.spawnSync("git", ["-C", root, "ls-tree", "-r", "--name-only", "HEAD"], {
    encoding: "utf8",
    timeout: 15000,
  });
  if (result.error || result.status !== 0) return null;
  return result.stdout.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
}

function scanSite(root) {
  if (!fs.existsSync(root)) {
    return { path: root, exists: false, html_files: [], asset_count: 0, chat_surface_candidates: [] };
  }
  const treeFiles = gitTreeFiles(root);
  const files = treeFiles || [];
  const htmlFiles = [];
  const chatCandidates = [];
  let assetCount = 0;

  if (treeFiles) {
    for (const rel of treeFiles) {
      const suffix = path.extname(rel).toLowerCase();
      if ([".html", ".htm"].includes(suffix)) htmlFiles.push(rel);
      if ([".png", ".jpg", ".jpeg", ".gif", ".svg", ".webp", ".css", ".js"].includes(suffix)) assetCount += 1;
      const lowered = rel.toLowerCase();
      if ([".html", ".htm", ".js", ".ts", ".tsx", ".jsx", ".json", ".md"].includes(suffix) &&
          chatSurfaceTokens.some((token) => lowered.includes(token))) {
        chatCandidates.push(rel);
      }
    }
  } else {
    function walk(current) {
      const entries = fs.readdirSync(current, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(current, entry.name);
        if (entry.isDirectory()) {
          walk(fullPath);
          continue;
        }
        const rel = path.relative(root, fullPath).split(path.sep).join("/");
        const suffix = path.extname(fullPath).toLowerCase();
        if ([".html", ".htm"].includes(suffix)) htmlFiles.push(rel);
        if ([".png", ".jpg", ".jpeg", ".gif", ".svg", ".webp", ".css", ".js"].includes(suffix)) assetCount += 1;
        const lowered = rel.toLowerCase();
        if ([".html", ".htm", ".js", ".ts", ".tsx", ".jsx", ".json", ".md"].includes(suffix) &&
            chatSurfaceTokens.some((token) => lowered.includes(token))) {
          chatCandidates.push(rel);
        }
      }
    }

    walk(root);
  }
  return {
    path: root,
    exists: true,
    html_files: htmlFiles.sort(),
    asset_count: assetCount,
    chat_surface_candidates: [...new Set(chatCandidates)].sort(),
    inventory_source: treeFiles ? "git-tree" : "filesystem",
  };
}

function scanReplica(root) {
  if (!fs.existsSync(root)) {
    return { path: root, exists: false, routes: [], components: [], chat_surface_candidates: [] };
  }

  const routes = [];
  const components = [];
  const chatCandidates = [];
  const treeFiles = gitTreeFiles(root);

  if (treeFiles) {
    for (const rel of treeFiles) {
      const suffix = path.extname(rel).toLowerCase();
      if (![".tsx", ".ts", ".jsx", ".js"].includes(suffix)) continue;
      const lowered = rel.toLowerCase();
      if (["route", "page", "app.tsx", "main.tsx"].some((token) => lowered.includes(token))) routes.push(rel);
      if (lowered.includes("component") || lowered.includes("/components/")) components.push(rel);
      if (chatSurfaceTokens.some((token) => lowered.includes(token))) chatCandidates.push(rel);
    }
  } else {
    function walk(current) {
      const entries = fs.readdirSync(current, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(current, entry.name);
        if (entry.isDirectory()) {
          walk(fullPath);
          continue;
        }
        const suffix = path.extname(fullPath).toLowerCase();
        if (![".tsx", ".ts", ".jsx", ".js"].includes(suffix)) continue;
        const rel = path.relative(root, fullPath).split(path.sep).join("/");
        const lowered = rel.toLowerCase();
        if (["route", "page", "app.tsx", "main.tsx"].some((token) => lowered.includes(token))) routes.push(rel);
        if (lowered.includes("component") || lowered.includes("/components/")) components.push(rel);
        if (chatSurfaceTokens.some((token) => lowered.includes(token))) chatCandidates.push(rel);
      }
    }

    walk(root);
  }

  return {
    path: root,
    exists: true,
    routes: [...new Set(routes)].sort(),
    components: [...new Set(components)].sort(),
    chat_surface_candidates: [...new Set(chatCandidates)].sort(),
    inventory_source: treeFiles ? "git-tree" : "filesystem",
  };
}

function buildPlan(site) {
  return site.html_files.map((htmlFile) => {
    let slug = htmlFile.replace(/\.html?$/, "");
    if (slug.endsWith("/index")) slug = slug.slice(0, -6);
    const route = slug === "" || slug === "index" ? "/" : `/${slug.replace(/^\/+|\/+$/g, "")}/`;
    return {
      route,
      source: htmlFile,
      status: "pending-migration",
      target: "prismtek-site-replica",
    };
  });
}

const args = parseArgs(process.argv.slice(2));
const site = scanSite(args.siteDir);
const replica = scanReplica(args.replicaDir);
const payload = {
  site,
  replica,
  discovery: {
    root: args.discoveryRoot,
    site_candidates: site.exists ? [] : discoverRepo(args.discoveryRoot, "prismtek-site"),
    replica_candidates: replica.exists ? [] : discoverRepo(args.discoveryRoot, "prismtek-site-replica"),
  },
  migration_plan: buildPlan(site),
  chat_agent_handoff: {
    website_owner_repo: "prismtek-site",
    runtime_contract_repo: "bmo-stack",
    site_candidates: site.chat_surface_candidates || [],
    replica_candidates: replica.chat_surface_candidates || [],
  },
};

if (!site.exists && payload.discovery.site_candidates.length) {
  payload.site.hint = "Use --site-dir with one of discovery.site_candidates.";
}
if (!replica.exists && payload.discovery.replica_candidates.length) {
  payload.replica.hint = "Use --replica-dir with one of discovery.replica_candidates.";
}

fs.mkdirSync(path.dirname(args.output), { recursive: true });
fs.writeFileSync(args.output, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
console.log(JSON.stringify(payload, null, 2));
